Ar validation

Tydik42 opened the pull request:
### New AR Verification Functionality:

*   **`core/algorithms/association_rules/ar_verifier/ar_verifier.h`**: Introduced the `ARVerifier` class to handle AR verification, including methods for loading data, executing the verification, and retrieving various statistics.
*   **`core/algorithms/association_rules/ar_verifier/ar_stats_calculator.h`**: Added the `ARStatsCalculator` class to compute actual support, confidence, and analyze violating rule clusters.
*   **`core/algorithms/association_rules/model/rule_coverage.h`**: Defined the `RuleCoverage` class to calculate the proportion of a rule within a transaction, used for cluster prioritization.
*   **`core/algorithms/association_rules/ar_verifier/enums.h`**: Added `ClusterPriority` enums to define different strategies for cluster analysis.

### Core Refactoring:

*   **`core/config/`**: Refactored shared options (`min_support`, `min_confidence`) into reusable `CommonOption` objects with strict `[0, 1]` validation to reduce code duplication between the AR miner and verifier.
*   **`core/model/transactional_data.h`**: Improved `TransactionalData` creation by introducing a dedicated `Params` struct.

vs9h commented on src/tests/test_ar_verifier.cpp:
```diff
@@ -0,0 +1,64 @@
+  ARVerifierRowsDataset, TestARVerifying,
+  ::testing::Values(
+      // test cases tabular
+     ARVerifyingParams(kRulesKaggleRows, {"MILK"}, {"BREAD"}, 0.2, 0.1, algos::InputFormat::tabular, 0, 0, 1),
```
Also, I don't like passing a large number of numbers, which makes the tests unreadable. It is necessary to compare which number is in which position. This can be avoided by adding an auxiliary structure and using aggregate initialization.

Here is a small sketch of the solution:
```cpp
struct ARVerifyingParams {
    algos::StdParamsMap params;

    struct TabularCSV {
        CSVConfig csv_config;
        unsigned int tid_col;
        unsigned int item_col;
    };

    struct Config {
        std::list<std::string> left;
        std::list<std::string> right;
        double min_sup;
        double min_conf;
    };

    ARVerifyingParams(TabularCSV const& table, Config config)
      : params({{onam::kCsvConfig, table.csv_config},
                {onam::kInputFormat, +algos::InputFormat::singular},
                {onam::kMinimumSupport, config.min_sup},
                {onam::kMinimumConfidence, config.min_conf},
                {onam::kARuleLeft, std::move(config.left)},
                {onam::kARuleRight, std::move(config.right)},
                {onam::kTIdColumnIndex, table.tid_col},
                {onam::kItemColumnIndex, table.item_col}}) {}

};
```

And write here something like this:
```cpp
     ARVerifyingParams(kRulesKaggleRowsSingular, {.left = {"MILK"}, .right ={"BREAD"}, .min_sup = 0.2, .min_conf = 0.1})
```

This will be much more readable and visual.

If you need more examples, I've done something similar before (`test_ind_verifier.c`, you can see here how the code is organized and do something similar) ​​and I've definitely promoted similar solutions for other primitives a few times (I don't remember which ones specifically).

vs9h reviewed (changes requested):
I partially looked at the pull request, I'll look at some parts in more detail later.

vs9h commented on src/core/algorithms/association_rules/ar_verifier/ar_verifier.cpp:
```diff
@@ -0,0 +1,112 @@
+unsigned long long ARVerifier::ExecuteInternal() {
+    std::vector<std::string> const& item_names_map = transactional_data_->GetItemUniverse();
+
+    std::vector<unsigned> ar_left_id;
```
It seems more correct to use `model::ColumnIndex`. If this is true, then it should be here and in those places where we refer to the index inside the rule.

vs9h commented on src/core/algorithms/association_rules/ar_verifier/ar_verifier.h:
```diff
@@ -0,0 +1,83 @@
+    bool first_column_tid_;
+
+    std::shared_ptr<model::TransactionalData> transactional_data_;
+    std::list<std::string> string_rule_left_;
+    std::list<std::string> string_rule_right_;
```
Why do we want to use a `std::list<>` here?

vs9h replied:
Can we get rid of using these lists? I think it would be better to replace them with a vector both in the existing AR code (if it exists) and in the new code.

vs9h commented on src/core/algorithms/association_rules/ar_verifier/ar_stats_calculator.cpp:
```diff
@@ -0,0 +1,98 @@
+
+void ARStatsCalculator::CalculateRuleCoverageCoefficients() {
+    for (auto const& [id, itemset] : data_->GetTransactions()) {
+        double coverage_left = CalculateTransactionCoverage(itemset.GetItemsIDs(), rule_.left);
+        double coverage_right = CalculateTransactionCoverage(itemset.GetItemsIDs(), rule_.right);
```
nit: I see that you use `const` for local variables. But in some places (like this one) you don't write them. I suggest adding them in all places for consistency here and in similar places. (This is not necessary)

ol-imorozko commented:
Take a look at https://github.com/Desbordante/desbordante-core/wiki/Development#оформление-коммитов -- there are guidelines on how we aim to write commit messages. 
I can see that your commits do not use imperative mood (which is an easy fix) and there aren't any descriptions on any of the commit messages. That's not necessarily a bad thing since usually there's no description needed anyway, but  that is something to note in the future

ol-imorozko commented:
Also I would suggest to avoid commits like "fix things". If a commit fixes something that was introduced in the previous comment, it's better to just fixup those changes. This way we will be left with one good commit instead of a flawed commit and a commit that fixes that.
Again, there are cases where this is acceptable. For example, a commit like `Apply clang-format` is generally acceptable.
But commits like [fix(AR): fixed bindings after merging main/origin branch](https://github.com/Desbordante/desbordante-core/pull/530/commits/e61fced1b86c66ee987ac391545c4a76d86bcc0f)
[e61fced](https://github.com/Desbordante/desbordante-core/pull/530/commits/e61fced1b86c66ee987ac391545c4a76d86bcc0f)
and
[fix(AR): fixed bindings (again)](https://github.com/Desbordante/desbordante-core/pull/530/commits/23a3843aedf3b578395ab25c2d870b0335bb83b5)
are not -- you should just write one good commit that adds bindings

vs9h commented on src/core/algorithms/association_rules/ar_verifier/ar_verifier.cpp:
```diff
@@ -0,0 +1,96 @@
+        check_rule_part(string_rule_part);
+    };
+
+    auto opts = config::TransactionalDataOptions(&transactional_data_params_);
+    RegisterOption(std::move(opts.input_table));
+    RegisterOption(std::move(opts.input_format));
+    RegisterOption(std::move(opts.first_column_tid));
+    RegisterOption(std::move(opts.item_column));
+    RegisterOption(std::move(opts.tid_column));
```
I'd prefer to change this approach quite a bit.

In an ideal world, we’d have the concept of complex options, so that registration could look like this:
```cpp
/* inside class: */
TransactionalData::Params data_params_;

/* here: */
RegisterOption(config::kTransactionalDataParams(&data_params_));
```

Inside `kTransactionalDataParams`, we would then register all related options. Currently, however, we don't have this functionality. I see two possible approaches:
1) Add support for such "complex" options.
2) Define the options in "config/transactional_data/" directory, similar to how other options are defined (e.g., the number of threads). Ideologically, these options all belong together, so I'd put them all in three files (`config/transactional_data/{types.h,options.h,options.cpp}`). In this case, each option would still be registered individually, for example:
```cpp
RegisterOption(config::kInputFormat(&data_params_.input_format)); 
RegisterOption(config::k...(&data_params_....)); 
...
```

I would say that any option would be adequate. As for the proposed variant with `std::move`, I'm not really in favor of it. It doesn't quite align with the current way we typically work with options.

vs9h commented on src/core/algorithms/association_rules/ar_verifier/ar_verifier.cpp:
```diff
@@ -0,0 +1,96 @@
+    RegisterOption(std::move(opts.item_column));
+    RegisterOption(std::move(opts.tid_column));
+
+    RegisterOption(config::kMinimumConfidenceOpt(&minconf_));
+    RegisterOption(config::kMinimumSupportOpt(&minconf_));
```
```suggestion
    RegisterOption(config::kMinimumSupportOpt(&minsup_));
```

vs9h replied:
Let's define options right here, because currently we use them only here (like for `kARuleLeft`, `kARuleRight`).

vs9h commented on src/core/algorithms/association_rules/ar_verifier/ar_verifier.cpp:
```diff
@@ -0,0 +1,96 @@
+#include "ar_verifier.h"
+
+#include <chrono>
```
unused

vs9h commented on src/core/algorithms/association_rules/ar_verifier/util/rule_coverage.h:
```diff
@@ -0,0 +1,64 @@
+
+namespace algos::ar_verifier::util {
+
+struct RuleCoverage {
```
Should we use class here?

Tydik42 replied:
The main point of `RuleCoverage` is to store two doubles, and all the methods are just convenient ways to check or evaluate them. So I thought it would be better to call it a structure.

vs9h commented on src/core/algorithms/association_rules/ar_verifier/util/rule_coverage.h:
```diff
@@ -0,0 +1,64 @@
+          right_coverage(CalculateCoverage(transaction_indices, right_rule_part)) {}
+
+private:
+    static double CalculateCoverage(std::vector<unsigned> const& transaction_indices,
```
Move to the source file?

vs9h commented on src/core/algorithms/association_rules/ar.h:
```diff
@@ -36,8 +54,8 @@ struct ARStrings {
 
-    ARStrings(std::list<std::string> left, std::list<std::string> right, double confidence,
-              double support)
+    ARStrings(std::list<std::string> left, std::list<std::string> right, double const confidence,
+              double const support)
```
Usually we don't use const for params, could you leave it as it was?

vs9h commented on src/core/config/ar_data/transactional_data_config.h:
```diff
@@ -0,0 +1,62 @@
+
+namespace config {
+
+struct TransactionalDataParams {
```
I'd prefer to put this structure right to the `TransactionalData` class:
```cpp
class TransactionalData {
public:
    struct Params {
        config::InputTable input_table;
        algos::InputFormat input_format;
        model::ColumnIndex tid_column_index;
        model::ColumnIndex item_column_index;
        bool first_column_tid;
    };
```

And add `CreateFrom` method, which takes this structure as param (existing `TransactionalData::CreateFrom` you can make private to prohibit their usage).

You can also pay attention to the fact that from the point of view of the language, it is strange that the parameters object has a `CreateFrom` method.

vs9h commented on src/python_bindings/ar/bind_ar_verification.h:
```diff
@@ -0,0 +1,7 @@
+#include <pybind11/pybind11.h>
+
+namespace python_bindings {
+void BindArVerification(pybind11::module_& main_module);
```
Since we use `AR` in `ARAlgorithm`, let's use everywhere the same naming convention:
```suggestion
void BindARVerification(pybind11::module_& main_module);
```

vs9h commented on src/tests/unit/test_apriori.cpp:
```diff
@@ -168,7 +168,7 @@ TEST_F(ARAlgorithmTest, PresentationDataset) {
 
 TEST_F(ARAlgorithmTest, SynteticDatasetWithPruning) {
-    auto algorithm = CreateAlgorithmInstance(kRulesSynthetic2, 0.13, 1.00001, 0, 1);
+    auto algorithm = CreateAlgorithmInstance(kRulesSynthetic2, 0.13, 1, 0, 1);
```
Why we must change it?

vs9h reviewed (changes requested):
Sorry for the very long review delay. I'll try to provide feedback as quickly as possible after the next round of changes.

If anything is unclear, please don't hesitate to ask.

Also, it would be great to split the history into separate commits (see the comments above from @ol-imorozko)

vs9h commented on src/tests/unit/test_ar_verifier.cpp:
```diff
@@ -0,0 +1,143 @@
+TEST_F(TestARVerifierError, TestInvalidRule) {
+    SingularDataset dataset{kRulesBook, 0, 1};
+    RuleConfig config{{"42"}, {"Milk"}, 0.1, 0.1};
+    ARVerifierParams params{dataset, config};
```
Let's use `kRulesBookSingular`

vs9h commented on src/tests/unit/test_ar_verifier.cpp:
```diff
@@ -0,0 +1,143 @@
+SingularDataset const kRulesBookSingular{kRulesBook, 0, 1};
+SingularDataset const kRulesPresentationSingular{kRulesPresentation, 0, 1};
+SingularDataset const kRulesPresentationExtSingular{kRulesPresentationExtended, 0, 1};
+TabularDataset const kRulesKaggleRowsTabular{kRulesKaggleRows, true};
```
Let's use Aggregate initialization here too.

vs9h commented on src/core/config/ar_data_threshold/option.h:
Let's split it into separate files (`config/min_confidence`, `config/min_support`).

And let's avoid using `ArThresholdType`. It's better to define a separate type for each option. And use those types in algorithms and in tests.

Also it's interesting, that we already have a minimum support option in `HyMD` algorithm. But in this case we use a number of pairs (`size_t` instead of `double`). So I suggest using `Ar` prefix for both type and option (`kArMinimum...`).

vs9h reviewed (changes requested):
Please, rewrite the commit history, this was mentioned above. 

It's better to get into the habit of writing a good commit history right away, and it's much more convenient to review PR in commit order.

I would divide the commits into options refactoring, ar verifier class implementation, examples adding, etc.

It would also be great to have commit descriptions.

vs9h commented on src/python_bindings/py_util/get_py_type.cpp:
```diff
@@ -101,7 +103,9 @@ py::tuple GetPyType(std::type_index type_index) {
-    };
+            PyTypePair<std::list<std::string>, kPyList, kPyStr>,
+            PyTypePair<std::unordered_map<std::string, std::vector<unsigned int>>, kPyDict, kPyStr,
+                       kPyList, kPyInt>};
```
Put here comma at the end (and in the file `py_to_any.cpp‎`)

vs9h commented on src/core/algorithms/association_rules/ar_verifier/ar_verifier.h:
```diff
@@ -0,0 +1,84 @@
+    ARStatsCalculator stats_calculator_;
+
+    double minsup_ = 0.0;
+    double minconf_ = 0.0;
```
use type aliases for these options

vs9h reviewed (changes requested):
I left a few comments.

Also, it's worth reorganizing the files inside `association_rules` folder. Firstly, I would rename it to `ar` (for consistency). Secondly, it's worth creating a separate folder for `apriori` algorithm and for `ar_verifier`.

Tydik42 committed:
Improve and unify configuration for AR and data models

- Standardize AR option names with "Ar" prefix to avoid conflicts.
- Centralize min_support/min_confidence into shared CommonOption
objects.
- Refactor TransactionalData to use a Params struct for cleaner
instantiation.
- Add strict [0, 1] range validation for support/confidence options.
- Switch AR rule storage from std::list to std::vector.
- Adapt AR miner and bindings to all related changes.
- Rename association_rulses to ar for consistency

Tydik42 committed:
Implement core logic for AR Verifier

- Add ARVerifier as the main algorithm interface.
- Add ARStatsCalculator for calculating sup, conf, and cluster metrics.
- Add RuleCoverage model and ClusterPriority enums as dependencies.
- Register the new algorithm within the application's framework.

Tydik42 committed:
Add bindings and tests for AR Verifier

Adds Python bindings for the AR Verifier algorithm. Introduces an
initial
test suite covering successful verification scenarios and edge cases.

Tydik42 committed:
Add example script and dataset for AR Verifier

Adds a Python example demonstrating how to use the AR Verifier with both
tabular and singular data.

vs9h commented on desbordante_run.log:
There are some, which should be removed

vs9h reviewed (commented):
There are very few new comments, let's fix them and try to merge this PR

Tydik42 commented on examples/basic/verifying_ar.py:
```diff
@@ -0,0 +1,172 @@
+          "the transaction #2 should contain 'Butter'. We will fix this record and re-validate the rule.")
+
+    algo = desbordante.ar_verification.algorithms.Default()
+    algo.load_data(table=(TABLE_FIXED_TABULAR, ',', False), input_format='tabular')
```
The CSV file has no header row — the first line is actual data

vs9h commented on src/core/algorithms/ucc/ucc_verifier/CMakeLists.txt:
```diff
@@ -1,4 +1,6 @@
 set(NAME ucc.verifier)
 desbordante_add_lib(NAME)
 target_sources(${NAME} PRIVATE ucc_verifier.cpp)
-target_link_libraries(${NAME} PRIVATE better-enums Boost::headers)
```
Why do we need to change this line?

Tydik42 replied:
It was necessary to fix a linker error caused by ar.verifier depending on model::transaction. Without this, `ColumnLayoutRelationData::CreateFrom`
and `PositionListIndex::Intersect` were unresolved at link time for ucc.verifier

vs9h commented on src/core/model/transaction/CMakeLists.txt:
```diff
@@ -1,3 +1,5 @@
 desbordante_add_lib(NAME OBJECT)
 target_sources(${NAME} PRIVATE itemset.cpp transactional_data.cpp)
+target_link_libraries(${NAME} PRIVATE better-enums)
+
```
Can you remove extra line here and in [‎src/core/config/CMakeLists.txt](https://github.com/Desbordante/desbordante-core/pull/530/changes#diff-1e71f9fb9a6c39d452fc1d333ef204c6f8b821f12534c3a4dfb6ca6c2716f2b9), [src/core/algorithms/ar/CMakeLists.txt](https://github.com/Desbordante/desbordante-core/pull/530/changes#diff-ef5a36701c122470ee850d06eb2f29d8e2b884739bedc6509addaa8583e32c67)?

vs9h commented on src/core/algorithms/ar/CMakeLists.txt:
```diff
@@ -1,7 +1,10 @@
 set(NAME ar)
 desbordante_add_lib(NAME)
-target_sources(${NAME} PRIVATE apriori.cpp ar_algorithm.cpp candidate_hash_tree.cpp)
+target_sources(${NAME} PRIVATE apriori/apriori.cpp ar_algorithm.cpp apriori/candidate_hash_tree.cpp)
```
It looks like we need add cmakelists file for apriori as well
